cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1240,11 +1250,141 @@ public long getProgramRunCount(ProgramId programId) {
     return result;
   }
 
+  void createStartTimeIfNotExist() {
+    MDSKey mdsKey = new MDSKey(RUN_COUNT_FIRST_UPGRADE_TIME);
+    if (get(mdsKey, Long.TYPE) == null) {
+      LOG.info("Start to upgrade the run count");
+      write(mdsKey, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()));
+    }
+  }
+
+  /**
+   * Compute the old run count if the run record is older than first start up time after 5.1
+   */
+  boolean computeOldRunCount(int maxRows) {
+    boolean upgradeDone = upgradeActiveRowkey(TYPE_RUN_RECORD_STARTING, maxRows);
```

Comment:
This can potentially do 2 * maxRows - 1 in one call to computeOldRunCount().

Also, if a bunch of calls are made to thie method because computeOldCompletedRunCount() needs to be called a lot, it will do a lot of unnecessary scans trying to upgrade old active rowkeys.

The caller should be handling this logic, moving from one 'phase' of the upgrade to the next.

Reply:
I separate the logic with upgrade active run records and doing the run count, and let the caller track the status of each upgrade